### PR TITLE
Deprecate `optuna.type_checking` module.

### LINKED ISSUE #1067: Drop Python 3.5 support
Thread to discussion dropping Python 3.5 support.

See also [Drop Python 2 support. #713]( https://github.com/optuna/optuna/issues/713)

### PR DESCRIPTION
**Pull request by c-bata**
## Motivation

To close #1067.

## Description of the changes

Deprecate `optuna.type_checking` module.

### CODE REVIEW

**Review by hvy — CHANGES_REQUESTED** [review]
Thank you for the PR! I left some comments but it otherwise LGTM.

File: optuna/__init__.py [hunk 1]
```diff
@@ -23,7 +24,6 @@
 from optuna.study import Study  # NOQA
 from optuna.trial import create_trial  # NOQA
 from optuna.trial import Trial  # NOQA
-from optuna.type_checking import TYPE_CHECKING  # NOQA
```
[the remark this answers is not in the document]
Comment (c-bata) [reply optuna/__init__.py:26]: I directly import `typing.TYPE_CHECKING` at L4. So seems not to break the user code. What do you think?

[the remark this answers is not in the document]
Comment (hvy) [reply optuna/__init__.py:26]: Ah, that case is fine but I meant something like the following.
```python
import optuna
if optuna.type_checking.TYPE_CHECKING: ...
```

[the remark this answers is not in the document]
Comment (c-bata) [reply optuna/__init__.py:26]: Ah, I see! I pushed baf3baf9. Thank you for your review!

**Review by hvy — APPROVED** [review]
LGTM! Thanks for your quick fixes.

I noticed we didn't put the version of removal in the message but I guess we can follow it up if deemed necessary. Current changes are at least consistent with `optuna.structs`.

**Comment by hvy** [comment]
Sorry could you fix the conflict?

File: optuna/__init__.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import importlib
 import types
 from typing import Any
+from typing import TYPE_CHECKING  # NOQA
```
[the remark this answers is not in the document]
Comment (c-bata) [reply optuna/__init__.py:4]: Right! And now `__all__` is defined at https://github.com/optuna/optuna/pull/2013. So we can obviously remove noqa comment.
Thank you for your review!

**Comment by himkt** [comment]
Thank you for the quick update, LGTM! 😸

**Review by himkt — APPROVED** [review]

**Pull request merged**